projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+
+  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+  private nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+  private nodeCoOrdinatesMap: Map<TopologyNode, TopologyCoordinates> = new Map<TopologyNode, TopologyCoordinates>();
+
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    this.findRootNodes(topology);
+    this.fillNodeAndLevelMaps();
+    this.fillNodeCoOrdinatesMap();
+    this.setNodeCoOrdinates(topology);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        this.nodeToLevelMap.set(node, 0);
+        this.levelToNodesMap.has(0) ? this.levelToNodesMap.get(0)?.push(node) : this.levelToNodesMap.set(0, [node])
```

Comment:
nit: Initialize levelToNodesMap before iterating, then simplify the logic below. E.g.
```ts
  this.levelNodesMap.set(0, []);

  ...

  this.levelToNodesMap.get(0)!.push(node); // Prefer the non null assertion to avoid hiding a logic error
```

It also might be helpful to use methods like getRootNodes rather than access things by index (which we do on line 28 too) to indicate the meaning to the reader.

Reply:
agreed, will update